fortanix-sgx-abi/src/lib.rs

```diff
@@ -611,18 +615,36 @@ impl Usercalls {
 /// concurrent usercalls with the same `id`, but it may reuse an `id` once the
 /// original usercall with that `id` has returned.
 ///
+/// An optional third queue can be used to cancel usercalls. To cancel an async
+/// usercall, the enclave should send the usercall's id and number on this
+/// queue. If the usercall has already been processed, the enclave may still
+/// receive a successful result for the usercall. Otherwise, the userspace will
+/// cancel the usercall's execution and return an [`Interrupted`] error on the
+/// return queue to notify the enclave of the cancellation. Note that usercalls
+/// that do not return [`Result`] cannot be cancelled and if the enclave sends
+/// a cancellation for such a usercall, the userspace should simply ignore it.
+/// Additionally, userspace may choose to ignore cancellations for non-blocking
+/// usercalls. Userspace should be able to cancel a usercall that has been sent
+/// by the enclave but not yet received by the userspace, i.e. if cancellation
+/// is received before the usercall itself. However, userspace should not keep
+/// cancellations forever since that would prevent the enclave from re-using
+/// usercall ids.
```

Comment:
This should be more concrete

Reply:
For example, you can synchronize the cancel queue with the usercall queue to maintain the following invariant: Whenever the enclave writes an id to the usercall or cancel queue, the enclave will not reuse that id until the usercall queue read pointer has advanced to the write pointer at the time the id was written. The corollary of this is that userspace can forget about an “unused” cancel id once the read pointer has advanced to a certain point. It may be possible to relax the invariant a bit to get less strict synchronization requirements but still have a decent bound on id liveness.

Reply:
That's an interesting idea. Implementing it would require some kind of strictly increasing read/write pointers that could be compared to each other. Here is one way to do it:
- Store an additional atomic "epoch" counter in the Receiver that records how many times the write offset of the ring buffer has wrapped around,
- Combine that value with the current read/write offset (modulo queue length) to form a read/write pointer.

To avoid problems when the produced numbers wrap around, we should make the "epoch" counter big enough and 64 bits seems plenty.

Can you think of a simpler solution?

Reply:
Can you update the documentation?